truth_justice_cake_cutting.py

```diff
@@ -0,0 +1,340 @@
+logger = logging.getLogger(__name__)
+    
+
+
+
+"""
+Get the intersection of two intervals
+"""
+def intervals_intersection(a,b):
+    intervals = []
+    for i in a:
+        for j in b:
+            interval = (max(i[0], j[0]), min(i[1], j[1]))
+            if(not interval[0]>=interval[1]):
+                intervals.append(interval)
+                
+    return intervals
+
+"""
+Get first element
+"""
+def first_element(i):
+    return i[0]
+
+"""
+Net implementation
+"""
+class Edge(object):
+  def __init__(self, u, v, w):
+    self.source = u
+    self.target = v
+    self.capacity = w
+
+  def __repr__(self):
+    return "%s->%s:%s" % (self.source, self.target, self.capacity)
+
+
+class FlowNetwork(object):
+  def  __init__(self):
+    self.adj = {}
+    self.flow = {}
+
+  def AddVertex(self, vertex):
+    self.adj[vertex] = []
+
+  def GetEdges(self, v):
+    return self.adj[v]
+
+  def AddEdge(self, u, v, w = 0):
+    if u == v:
+      raise ValueError("u == v")
+    edge = Edge(u, v, w)
+    redge = Edge(v, u, 0)
+    edge.redge = redge
+    redge.redge = edge
+    self.adj[u].append(edge)
+    self.adj[v].append(redge)
+    # Intialize all flows to zero
+    self.flow[edge] = 0
+    self.flow[redge] = 0
+    
+  def FindPath(self, source, target, path):
+    if source == target:
+      return path
+    for edge in self.GetEdges(source):
+      residual = edge.capacity - self.flow[edge]
+      if residual > 0 and not (edge, residual) in path:
+        result = self.FindPath(edge.target, target, path + [(edge, residual)])
+        if result != None:
+          return result
+
+  def MaxFlow(self, source, target):
+    path = self.FindPath(source, target, [])
+    while path != None:
+      flow = min(res for edge, res in path)
+      for edge, res in path:
+        self.flow[edge] += flow
+        self.flow[edge.redge] -= flow
+      path = self.FindPath(source, target, [])
+    return sum(self.flow[edge] for edge in self.GetEdges(source))
+
+
+
+
+"""
+agents: list of PiecewiseLinearAgent
+"""
+def algorithm2(agents):
+    """
+    >>> a = PiecewiseLinearAgent([(0,0.39, 1,1)], "a")
+    >>> b = PiecewiseLinearAgent([(0,0.6, 1,1)], "b")
+    >>> c = PiecewiseLinearAgent([(0,0.1, 1,1)], "c")
+    >>> agents = [a,b,c]
+    
+    >>> allocations = algorithm2(agents)
+    
+    >>> print(allocations)
+    
+    algorithm2
+    > a gets [(0.0, 0.2), (1.0, 1.2000000000000002), (0.2, 0.3), (0.7, 0.7999999999999998)] with value 2.30
+    > b gets [(0.2, 0.4), (0.8, 1.0), (0.3, 0.39999999999999997), (0.5999999999999999, 0.7)] with value 2.55
+    """
+    print("algorithm2")
+    N = len(agents)
+    logger.info("N is: %d", N)
+    
+    #organise the intervals
+    intervals = []
+    for agent in agents:
+        regions = agent.desired_regions
+        for region in regions:
+            intervals.append((region[0],'start'))
+            intervals.append((region[1],'end'))
+    intervals = sorted(intervals, key = first_element)
+    
```

Comment:
Please add more "logger.info" messages, to explain better how the algorithm works.